src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java

```diff
@@ -2,42 +2,84 @@
+import com.fasterxml.jackson.databind.exc.InvalidFormatException;
+import com.fasterxml.jackson.databind.exc.MismatchedInputException;
 import java.util.HashMap;
+import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.converter.HttpMessageNotReadableException;
 import org.springframework.web.bind.annotation.ControllerAdvice;
-import org.springframework.web.bind.annotation.ExceptionHandler;
-import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.context.request.WebRequest;
 import org.springframework.web.servlet.NoHandlerFoundException;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import uk.gov.companieshouse.api.accounts.model.validation.Error;
+import uk.gov.companieshouse.api.accounts.model.validation.Errors;
+import uk.gov.companieshouse.api.accounts.validation.ErrorMessageKeys;
+import uk.gov.companieshouse.api.accounts.validation.ErrorType;
+import uk.gov.companieshouse.api.accounts.validation.LocationType;
 import uk.gov.companieshouse.logging.Logger;
 import uk.gov.companieshouse.logging.LoggerFactory;
 
-
 /**
  * GlobalExceptionHandler defines handlers for generic exceptions.
  *
  * Api Specific Errors are handled in the Controller.
  */
 @ControllerAdvice
-public class GlobalExceptionHandler {
-
-  private static final Logger LOGGER = LoggerFactory.getLogger(APPLICATION_NAME_SPACE);
-
-  @ExceptionHandler(value = {NoHandlerFoundException.class})
-  @ResponseStatus(value = HttpStatus.NOT_FOUND)
-  protected void handleNoHandlerFoundException(Exception ex) {
-    logError(ex);
-  }
-
-  @ExceptionHandler(value = {RuntimeException.class, Exception.class})
-  @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
-  protected void handleException(Exception ex) {
-    logError(ex);
-  }
-
-  private void logError(Exception ex) {
-    HashMap<String, Object> message = new HashMap<>();
-    message.put("message", ex.getMessage());
-    message.put("error", ex.getClass());
-    LOGGER.error(ex, message);
-  }
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+    private static final Logger STRUCTURED_LOGGER = LoggerFactory.getLogger(APPLICATION_NAME_SPACE);
+
+    @Override
+    protected ResponseEntity<Object> handleNoHandlerFoundException(NoHandlerFoundException ex,
+        HttpHeaders headers, HttpStatus status, WebRequest request) {
+        logClientError(ex);
+        return super.handleNoHandlerFoundException(ex, headers, status, request);
+    }
+
+    @Override
+    protected ResponseEntity<Object> handleHttpMessageNotReadable(
+        HttpMessageNotReadableException ex, HttpHeaders headers, HttpStatus status,
+        WebRequest request) {
+        logClientError(ex);
+
+        StringBuilder message = new StringBuilder("JSON parse exception");
```

Comment:
sorry to be that guy, but is there another way around the potential if else continuum?